### PR TITLE
[SPARK-20393][WEBU UI][1.6] Strengthen Spark to prevent XSS vulnerabilities

### PR DESCRIPTION
**Pull request by ambauma**
## What changes were proposed in this pull request?

This is the fix for the master branch applied to the 1.6 branch.  My (unnamed) company will be using Spark 1.6 probably for another year.   We have been blocked from having Spark 1.6 on our workstations until CVE-2017-7678 is patched, which SPARK-20393 does.  I realize there will not be an official Spark 1.6.4 release, but it still seems wise to keep the code there patched for those who are stuck on that version.  Otherwise I imagine several forks duplicating 1.6 compliance and security fixes.

## How was this patch tested?

The patch came with unit tests.  The test build passed.  Manual testing on one of the effected screens showed the newline character removed.  Screen display was the same regardless (html ignores newline characters).  
screenshot from 2017-10-18 14-18-17

Please review http://spark.apache.org/contributing.html before opening a pull request.

The patch itself is from previous pull requests associated to SPARK-20939.  My original "work" was actions on what to apply to branch 1.6. and I license the work to the project under the project’s open source license.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @ambauma .
Thank you for contributing.
- SPARK-20393 should be backported branch-2.0 first.
- PR titles should be the same with the original one except the version numbers.
```
[SPARK-20393][WEBU UI][2.0] Strengthen Spark to prevent XSS vulnerabilities
[SPARK-20393][WEBU UI][1.6] Strengthen Spark to prevent XSS vulnerabilities
```

**Comment by ambauma** [comment]
Understood.  Working on porting to 2.0...

**Comment by srowen** [comment]
I don't think there are any more 1.x releases coming, and doubt there are more 2.0.x releases. Do you really need this in Spark or is it something you can apply to your own release branch?

**Comment by ambauma** [comment]
I have a release in my fork for my immediate needs.  However, Spark 1.6 is still included in Hortonworks and is default in Cloudera.  This patch addresses CVE-2017-7678.  Some companies in strict regulatory environments may fail audits and be forced to remove Spark 1.6 if it is not patched.  Rather than keeping security patches in forks, I think it makes sense to merge them back into the mainline for branches that are still in active use.  That way if I get hit by a bus and CVE-2018-XXXX comes out, CVE-2017-7678 will already be covered and the work will not need to be duplicated.

**Comment by srowen** [comment]
(Spark 1.x is legacy in Cloudera, but, it has its own 1.x branch anyway)
I think it's not a big deal to backport if it goes into later branches first, sure. But I doubt there is another 1.x release here.

**Comment by ambauma** [comment]
I just posted the 2.0 pull request.  https://github.com/apache/spark/pull/19538

**Comment by ambauma** [comment]
Working on duplicating PySpark failures...

**Comment by ambauma** [comment]
Able to duplicate.  Working theory is that this is related to numpy 1.12.1.  Here is my conda env:
(spark-1.6) andrew@andrew-Inspiron-7559:~/git/spark$ conda list
# packages in environment at /home/andrew/.conda/envs/spark-1.6:
#
ca-certificates           2017.08.26           h1d4fec5_0  
certifi                   2016.2.28                py34_0  
intel-openmp              2018.0.0             h15fc484_7  
libedit                   3.1                  heed3624_0  
libffi                    3.2.1                h4deb6c0_3  
libgcc-ng                 7.2.0                h7cc24e2_2  
libgfortran               1.0                           0  
libstdcxx-ng              7.2.0                h7a57d05_2  
mkl                       2017.0.3                      0  
ncurses                   6.0                  h06874d7_1  
numpy                     1.12.1                   py34_0  
openblas                  0.2.19                        0  
openssl                   1.0.2l               h077ae2c_5  
pip                       9.0.1                    py34_1  
python                    3.4.5                         0  
readline                  6.2                           2  
setuptools                27.2.0                   py34_0  
sqlite                    3.13.0                        0  
tk                        8.5.18                        0  
wheel                     0.29.0                   py34_0  
xz                        5.2.3                h2bcbf08_1  
zlib                      1.2.11               hfbfcf68_1

**Comment by ambauma** [comment]
Believed fixed.  Hard to say for sure without knowing the precise python and numpy versions the build is using.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala [hunk 1]
```diff
@@ -16,9 +16,9 @@
  */
 
 package org.apache.spark.ui.jobs
-
+import javax.servlet.http.HttpServletRequest
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala]: Hm, I'm not sure if this back-port is correct. This file's change doesn't look like it does anything and I don't see this change in the original: https://github.com/apache/spark/pull/17686/files

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/ui/DriverPage.scala [hunk 2]
```diff
@@ -0,0 +1,180 @@
+/*
```
Comment (srowen) [inline resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/ui/DriverPage.scala]: Likewise this isn't part of the backport is it? https://github.com/apache/spark/pull/17686/files

File: python/pyspark/mllib/classification.py [hunk 3]
```diff
@@ -173,7 +173,7 @@ def __init__(self, weights, intercept, numFeatures, numClasses):
             self._dataWithBiasSize = None
             self._weightsMatrix = None
         else:
-            self._dataWithBiasSize = self._coeff.size / (self._numClasses - 1)
+            self._dataWithBiasSize = self._coeff.size // (self._numClasses - 1)
```
Comment (srowen) [inline python/pyspark/mllib/classification.py]: Nor this?

**Review by srowen — CHANGES_REQUESTED** [review]

File: python/pyspark/mllib/classification.py (hunk 3 above)
Comment (ambauma) [reply python/pyspark/mllib/classification.py]: I had to apply this to get past a python unit test failure.  My assumption is that the NewSparkPullRequestBuilder is on a different version of numpy than when the Spark 1.6 branch was last built.  The current python unit test failure looks like it has to do with a novel version of SciPy.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala (hunk 1 above)
Comment (ambauma) [reply core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala]: Will look into this...

File: python/pyspark/mllib/classification.py (hunk 3 above)
Comment (ambauma) [reply python/pyspark/mllib/classification.py]: This is already fixed in the 2.0 branch, btw.  Just was never applied to 1.6.  [SPARK-20862]

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/ui/DriverPage.scala (hunk 2 above)
Comment (ambauma) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/ui/DriverPage.scala]: I'll look into this as well...

File: core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala (hunk 1 above)
Comment (ambauma) [reply core/src/main/scala/org/apache/spark/ui/jobs/JobsTab.scala]: Agreed, will remove.

File: python/pyspark/mllib/classification.py (hunk 3 above)
Comment (ambauma) [reply python/pyspark/mllib/classification.py]: The NewSparkPullRequestBuilder failed on python tests.  I was only able to duplicate the failure with Python 3.4 and numpy 1.12.1, which I'm guessing is the versions that NewSparkPullRequestBuilder is using.  Older and newer versions of numpy build clean either way.

File: resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/ui/DriverPage.scala (hunk 2 above)
Comment (ambauma) [reply resource-managers/mesos/src/main/scala/org/apache/spark/deploy/mesos/ui/DriverPage.scala]: I'm not sure what I did to make this whole file look new, but I've copied the 1.6 current and reapplied stripXSS locally.  Waiting for my build to pass to commit again.

**Comment by felixcheung** [comment]
Jenkins, retest this please

**Comment by ambauma** [comment]
I'm unable to duplicate the PySpark failures locally.  I assume I need a specific version of SciPy to duplicate the error.  Is there a way I could get what versions the build server is running?  Something like:
`sorted(["%s==%s" % (i.key, i.version) for i in pip.get_installed_distributions()])` for python and python 3.4?

**Comment by felixcheung** [comment]
@shaneknapp - could you help check - what version of SciPy Jenkins is running with? thanks!

**Comment by shaneknapp** [comment]
python2.7:  0.17.0
python3:  0.18.1

**Comment by shaneknapp** [comment]
ok to test

**Comment by felixcheung** [comment]
Jenkins test this please

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
retest this please

**Comment by felixcheung** [comment]
Jenkins test this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Pull request closed without merging**